File: Ray Tracing in One Weekend - Python/util/metal/hittable_list.py
from util.metal.hit_record import HitRecord
from util.interval import Interval


class HittableList:

    def __init__(self, obj=None):
        self.objects = []
        if obj is not None:
            self.add(obj)

    def add(self, obj):
        self.objects.append(obj)

    def hit(self, ray, ray_t, rec):

        # ==========================================
        # TEMP RECORD (used for comparisons)
        # ==========================================
        temp_rec = HitRecord()

        hit_anything = False

        # IMPORTANT FIX:
        # ray_t MUST be Interval (not float)
        closest_so_far = ray_t.max

        for obj in self.objects:

            # Shrink valid interval each hit
            interval = Interval(ray_t.min, closest_so_far)

            if obj.hit(ray, interval, temp_rec):

                hit_anything = True
                closest_so_far = temp_rec.t

                # copy valid hit into final record
                rec.copy(temp_rec)

        return hit_anything

File: Ray Tracing next Week - Javascript/Ch5-C/hittable_list.js
class HittableList extends Hittable {

    constructor(objects = []) {
        super();

        this.objects = objects;
        this.bbox = null;

        for (const object of this.objects) {
            if (this.bbox === null)
                this.bbox = object.bounding_box();
            else
                this.bbox = new AABB({
                    box0: this.bbox,
                    box1: object.bounding_box()
                });
        }
    }

    clear() {
        this.objects = [];
        this.bbox = null;
    }

    add(object) {
        this.objects.push(object);

        if (this.bbox === null)
            this.bbox = object.bounding_box();
        else
            this.bbox = new AABB({
                box0: this.bbox,
                box1: object.bounding_box()
            });
    }

    hit(r, ray_t, hit_record) {

        let temp_rec = new HitRecord();
        let hit_anything = false;
        let closest_so_far = ray_t.max;

        for (const object of this.objects) {

            if (object.hit(
                r,
                new Interval({ min: ray_t.min, max: closest_so_far }),
                (rec)=>{
                temp_rec = rec;
            }
            )) {

                hit_anything = true;
                closest_so_far = temp_rec.t;
                hit_record(temp_rec);
            }
        }

        return hit_anything;
    }

    bounding_box() {
        return this.bbox;
    }
}

window.HittableList = HittableList;